fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java

```diff
@@ -2587,24 +2590,46 @@ private static void manageException(String message, IssueType issueType, FHIRSea
-
-                                // Check that the target resource type of the reference matches one of the
-                                // target resource types in the compartment definition.
-                                final String compartmentName = rv.getTargetResourceType();
-                                if (paramEntry.getValue().contains(compartmentName)) {
-                                    // Add this reference to the set of references we're collecting for each compartment
-                                    CompartmentReference cref = new CompartmentReference(searchParm, compartmentName, rv.getValue());
-                                    Set<CompartmentReference> references = result.computeIfAbsent(compartmentName, k -> new HashSet<>());
-                                    references.add(cref);
+                            String compartmentName = null;
+                            String compartmentId = null;
+
+                            Element element = node.asElementNode().element();
+                            if (element.is(Reference.class)) {
+                                Reference reference = element.as(Reference.class);
+                                ReferenceValue rv = ReferenceUtil.createReferenceValueFrom(reference, baseUrl);
+                                if (rv.getType() != ReferenceType.DISPLAY_ONLY && rv.getType() != ReferenceType.INVALID) {
+                                    compartmentName = rv.getTargetResourceType();
+                                    compartmentId = rv.getValue();
+
+                                    // Check that the target resource type of the reference matches one of the
+                                    // target resource types in the compartment definition.
+                                    if (!paramEntry.getValue().contains(compartmentName)) {
+                                        if (log.isLoggable(Level.FINE)) {
+                                            log.fine("Skipping reference with value " + reference.getReference() + ";"
+                                                    + " target resource type does not match any of the allowed compartment types: " + paramEntry);
+                                        }
+                                        continue;
+                                    }
+                                }
+                            } else if (element.is(FHIR_STRING)) {
+                                if (paramEntry.getValue().size() != 1) {
+                                    log.warning("CompartmentDefinition inclusion criteria must be of type Reference unless they have 1 and only 1 resource target");
+                                    continue;
                                 }
+                                compartmentName = paramEntry.getValue().iterator().next();
+                                compartmentId = element.as(FHIR_STRING).getValue();
```

Comment:
previously, we supported only inclusion criteria that selected elements of type Reference.  now we will support elements of type String as well, but only in cases where the compartment type is unambiguous (so that we do not need to check the target resourceType from the reference to determine which compartment the value is for).